static_frame/core/index_hierarchy.py

```diff
@@ -351,40 +538,69 @@ def from_index_items(cls: tp.Type[IH],
             items: iterable of pairs of label, :obj:`Index`.
             index_constructor: Optionally provide index constructor for outermost index.
         '''
-        labels = []
-        index_levels = []
+        labels: tp.List[tp.Hashable] = []
+        index_inner: tp.Optional[IndexGO] = None  # We will grow this in-place
+        indexers_inner: tp.List[np.ndarray] = []
+
+        # Contains the len of the index for each label. Used to generate the outermost indexer
+        repeats: tp.List[int] = []
 
-        offset = 0
         for label, index in items:
             labels.append(label)
-            index = mutable_immutable_index_filter(cls.STATIC, index) #type: ignore
-            index_levels.append(cls._LEVEL_CONSTRUCTOR(
-                    index,
-                    offset=offset,
-                    own_index=True)
-            )
-            offset += len(index)
-
-        targets = ArrayGO(np.array(index_levels, dtype=object), own_iterable=True)
-
-        index_outer = index_from_optional_constructor(labels,
-                    default_constructor=cls._INDEX_CONSTRUCTOR,
-                    explicit_constructor=index_constructor)
-        levels = cls._LEVEL_CONSTRUCTOR(
-                index=index_outer, #type: ignore
-                targets=targets,
-                own_index=True,
-                depth_reference=2, # depth always 2
+
+            if index_inner is None:
+                # This is the first index. Create an IndexGO
+                index_inner = tp.cast(IndexGO, mutable_immutable_index_filter(False, index))
+                new_indexer = PositionsAllocator.get(len(index_inner))
+            else:
+                new_labels = index.difference(index_inner) # Retains order!
+
+                if new_labels.size:
+                    index_inner.extend(new_labels.values)
+
+                new_indexer = index.iter_label().apply(index_inner._loc_to_iloc)
+
+            indexers_inner.append(new_indexer)
+            repeats.append(len(index))
+
+        if not labels:
+            assert index_inner is None # sanity check
+            return cls._from_empty(EMPTY_TUPLE, name=name, depth_reference=2)
+
+        index_inner = mutable_immutable_index_filter(cls.STATIC, index_inner) # type: ignore
+        assert isinstance(index_inner, Index) # mypy
```

Comment:
I do not like adding `isinstance` for mypy... if the alternative is a `#type: ignore`, that is a fine option.

Reply:
Okay I will remove